### PR TITLE
Revamp cover art providers UI (new version)

### PR DESCRIPTION
**Pull request by zas**
- make cover art providers more modular
- providers can be re-ordered to change provider's excution order
- support for provider-specific options, displayed as sub-pages of the main cover art options page
- clean up the API

capture du 2015-07-03 18 47 21
capture du 2015-07-03 18 47 29

### CODE REVIEW

File: picard/ui/sortablecheckboxlist.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    def __repr__(self):
+        params = []
+        params.append('text=' + repr(self.text))
+        params.append('checked=' + repr(self.checked))
+        if self.data is not None:
+            params.append('data=' + repr(self.data))
+        return "%s(%s)" % (self.__class__.__name__, ", ".join(params))
+
+
+def main():
```
Comment (mineo) [inline picard/ui/sortablecheckboxlist.py]: This doesn't look like it should be here.

File: picard/ui/sortablecheckboxlist.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
+from PyQt4 import QtGui
+from PyQt4 import QtCore
+import sys
+from functools import partial
+
+
+class Signal:
+
+    """Signal class from http://blog.abstractfactory.io/dynamic-signals-in-pyqt/
```
Comment (mineo) [inline picard/ui/sortablecheckboxlist.py]: Could you describe why we can't use qts signals here?

File: picard/ui/sortablecheckboxlist.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        layout.addWidget(up_button, row, self._BUTTON_UP)
+        layout.addWidget(down_button, row, self._BUTTON_DOWN)
+        self.setSignals(row)
+
+    def _emit_changed(self):
+        if not self.__no_emit:
+            self.changed.emit(self.__items)
+
+
+class SortableCheckboxListItem:
```
Comment (mineo) [inline picard/ui/sortablecheckboxlist.py]: Most of our other classes seem to be new-style ones, maybe these should be as well, if only for consistency?

File: picard/coverart/providers/caa.py [hunk 4]
```diff
@@ -107,8 +247,8 @@ def _has_suitable_artwork(self):
 
     def enabled(self):
         """Check if CAA artwork has to be downloaded"""
-        if not config.setting['ca_provider_use_caa']:
-            log.debug("Cover Art Archive disabled by user")
+        enabled = CoverArtProvider.enabled(self)
+        if not enabled:
             return False
```
Comment (mineo) [inline picard/coverart/providers/caa.py]: Why not `return not CoverArtProvider.enabled(self)` if we're not doing any logging here?

File: picard/coverart/providers/caa.py [hunk 5]
```diff
@@ -36,6 +42,139 @@
...
+        self.setWindowTitle(_("Cover art types"))
+        self._items = {}
+        self.layout = QtGui.QVBoxLayout(self)
+
+        grid = QtGui.QWidget()
+        gridlayout = QtGui.QGridLayout()
+        grid.setLayout(gridlayout)
+
+        rows = len(CAA_TYPES) // self._columns + 1
+        positions = [(i, j) for i in range(rows) for j in range(self._columns)]
```
Comment (mineo) [inline picard/coverart/providers/caa.py]: This is highly subjective, but I'd find
```python
for index, caa_type in enumerate(CAA_TYPES):
    row = index // self._columns
    column = index % self._columns
    ...
    gridlayout.addWidget(item, row, column)
```
easier to read.

File: picard/ui/options/cover.py [hunk 6]
```diff
@@ -141,7 +59,31 @@ def __init__(self, parent=None):
...
         self.ui.save_images_to_files.clicked.connect(self.update_filename)
-        self.ui.restrict_images_types.clicked.connect(self.update_caa_types)
+
+    def updateProvidersOption(self, items):
+        new = []
+        for item in items:
+            new.append((item.data, item.checked))
+        config.setting['ca_providers'] = new
+
+    def load_ca_providers(self):
```
Comment (mineo) [inline picard/ui/options/cover.py]: https://github.com/musicbrainz/picard/pull/413/files#diff-5efe8de01fc9a85826ac03f4a048501aL30 does not shorten `cover_art`.

File: picard/ui/options/cover.py [hunk 7]
```diff
@@ -141,7 +59,31 @@ def __init__(self, parent=None):
...
+        """
+            Load available providers, initialize tabs, restore state of each
```
Comment (mineo) [inline picard/ui/options/cover.py]: From a quick grep, it looks like the comment style for methods on classes is to put the content immediately after `"""`.

File: picard/coverart/providers/__init__.py [hunk 8]
```diff
@@ -17,23 +17,74 @@
...
+from picard.ui.options import OptionsPage, register_options_page
 
 
 _cover_art_providers = ExtensionPoint()
 
 
+class ProviderOptions(OptionsPage):
+
+    """
+        Template class for provider's options
```
Comment (mineo) [inline picard/coverart/providers/__init__.py]: This documentation needs to explain how to make use of the template.

File: picard/coverart/providers/__init__.py [hunk 9]
```diff
@@ -17,23 +17,74 @@
...
+    def __init__(self, parent=None):
+        super(ProviderOptions, self).__init__(parent)
+        if callable(self._options_ui):
+            self.ui = self._options_ui()
+            self.ui.setupUi(self)
+
+    def load(self):
+        pass
+
+    def save(self):
```
Comment (mineo) [inline picard/coverart/providers/__init__.py]: `Active`, `load`, `save` are already provided with the same values by `OptionsPage`, and I think setting `SORT_ORDER` to 0 instead of 1000 (the value from `OptionsPage`) makes no difference, so these could be removed.

File: picard/coverart/providers/__init__.py [hunk 10]
```diff
@@ -17,23 +17,74 @@
...
+
+    """
+        Template class for provider's options
+    """
+
+    PARENT = "cover"
+    SORT_ORDER = 0
+    ACTIVE = True
+
+    options = []
```
Comment (mineo) [inline picard/coverart/providers/__init__.py]: This doesn't seem to be used anywhere.

File: picard/coverart/providers/__init__.py [hunk 11]
```diff
@@ -17,23 +17,74 @@
...
+
+    _options_ui = None
+
+    def __init__(self, parent=None):
+        super(ProviderOptions, self).__init__(parent)
+        if callable(self._options_ui):
```
Comment (mineo) [inline picard/coverart/providers/__init__.py]: If `_options_ui` is not callable the user/developer of a subclass of this has misused this class, that should definitely be an exception.

This also doesn't seem to be very specific to cover art provider options pages - all option pages need to initialize their UI somehow. Do we want to move this up to `OptionsPage`?

File: picard/coverart/providers/__init__.py [hunk 12]
```diff
@@ -17,23 +17,74 @@
...
     providers = []
-    for p in _cover_art_providers:
+    for p in sorted(_cover_art_providers, key=_key_provider):
         providers.append(p)
+    log.debug("CA Providers order: %s",
+              ' > '.join([p.NAME for p in providers]))
     return providers
 
 
+def is_provider_enabled(provider_name):
```
Comment (mineo) [inline picard/coverart/providers/__init__.py]: Neither `CoverArtProvider.enabled` and `is_provider_enabled` document how they differ, although both from their contexts they both seem to return whether a cover art provider is enabled.

This is only used when a provider object is already in scope, so this could be a method on `CoverArtProvider`.

File: picard/coverart/providers/__init__.py [hunk 13]
```diff
@@ -17,23 +17,74 @@
...
+    enabled = False
+    for name, checked in config.setting['ca_providers']:
+        if name == provider_name:
+            enabled = checked
```
Comment (mineo) [inline picard/coverart/providers/__init__.py]: Can't we shorten this to `return checked` here?

File: picard/ui/sortablecheckboxlist.py (hunk 2 above)
Comment (zas) [inline picard/ui/sortablecheckboxlist.py]: We could i guess, but i didn't want to bother with them.
I don't like pyqtsignals mostly because of the "typed" arguments.
Frankly, this little class does the job perfectly, is simpler to use and more pythonic.
Though it should be somewhere in util/

File: picard/coverart/providers/caa.py (hunk 4 above)
Comment (zas) [inline picard/coverart/providers/caa.py]: > Why not return not CoverArtProvider.enabled(self) if we're not doing any logging here?

It is logged, but the diff isn't very clear here ;)
https://github.com/zas/picard/blob/revamp_cover_providers_2/picard/coverart/providers/caa.py#L248-256

File: picard/coverart/providers/__init__.py (hunk 11 above)
Comment (zas) [inline picard/coverart/providers/__init__.py]: > Do we want to move this up to OptionsPage?

It would make sense, but not in this PR. I'll do it after merge.

File: picard/coverart/providers/__init__.py (hunk 12 above)
Comment (zas) [inline picard/coverart/providers/__init__.py]: Clarified in https://github.com/zas/picard/commit/2e66ff99608eed43dd009ddae55fb2ec8385fdb1

File: picard/ui/sortablecheckboxlist.py (hunk 1 above)
Comment (zas) [inline picard/ui/sortablecheckboxlist.py]: Removed in https://github.com/zas/picard/commit/f82056007cd3038f432c8450930ce80c95306e3b

File: picard/ui/sortablecheckboxlist.py (hunk 3 above)
Comment (zas) [inline picard/ui/sortablecheckboxlist.py]: Fixed in https://github.com/zas/picard/commit/7f2b8459a4d20256d53120a8416eec9ea9ede2d2

File: picard/coverart/providers/caa.py (hunk 5 above)
Comment (zas) [inline picard/coverart/providers/caa.py]: Agreed, modified in https://github.com/musicbrainz/picard/commit/951e08afd77e09235ff4901b791de2acfd14b148

File: picard/ui/options/cover.py (hunk 6 above)
Comment (zas) [inline picard/ui/options/cover.py]: Renamed in https://github.com/zas/picard/commit/2b9688bfdc4023993a891f5c289172682c854da0

File: picard/ui/options/cover.py (hunk 7 above)
Comment (zas) [inline picard/ui/options/cover.py]: https://github.com/zas/picard/commit/53a864e6ade2750f73451a89babab3639a00bd7f

File: picard/coverart/providers/__init__.py (hunk 8 above)
Comment (zas) [inline picard/coverart/providers/__init__.py]: https://github.com/zas/picard/commit/94381dacee2ef8f1cd8ebe008d4d4cbbb1484fa1

File: picard/coverart/providers/__init__.py (hunk 9 above)
Comment (zas) [inline picard/coverart/providers/__init__.py]: Yes, see https://github.com/zas/picard/commit/6ca42d241f655ee1d11eb276743b3f95f96199c1

File: picard/coverart/providers/__init__.py (hunk 10 above)
Comment (zas) [inline picard/coverart/providers/__init__.py]: Removed in https://github.com/zas/picard/commit/f5821ca4aa0d24deb21b275974a103654438d0ca

File: picard/coverart/providers/__init__.py (hunk 11 above)
Comment (zas) [inline picard/coverart/providers/__init__.py]: > If _options_ui is not callable the user/developer of a subclass of this has misused this class, that should definitely be an exception.

Agreed, fixed in https://github.com/zas/picard/commit/c4eaa3cb39deeff4781bdce546419f77ab6bddf8

File: picard/coverart/providers/__init__.py (hunk 13 above)
Comment (zas) [inline picard/coverart/providers/__init__.py]: Ahah, done before i saw this one: https://github.com/zas/picard/commit/51b7c5bf9df4764fc1f7f973ddbb2197e7cab8cb

File: picard/coverart/providers/__init__.py [hunk 14]
```diff
@@ -17,23 +17,80 @@
...
 
+class ProviderOptions(OptionsPage):
+
+    """ Template class for provider's options
+
+        It works like OptionsPage for the most (options, load, save)
+        It will append the provider's options page as a child of the main
+        cover art's options page.
+
+        The property _options_ui should be set to a valid Qt Ui class
```
Comment (mineo) [inline picard/coverart/providers/__init__.py]: "Should" usually means "this is recommended, but not necessary".

File: picard/ui/options/cover.py [hunk 15]
```diff
@@ -141,7 +59,28 @@ def __init__(self, parent=None):
...
+
+    def update_providers_options(self, items):
+        config.setting['ca_providers'] = [(item.data, item.checked)
+                                          for item in items]
+
+    def load_cover_art_providers(self):
+        """Load available providers, initialize tabs, restore state of each
+        """
+        widget = SortableCheckboxListWidget()
+        self.provider_options = defaultdict(list)
```
Comment (mineo) [inline picard/ui/options/cover.py]: Is this used somewhere else?

File: picard/coverart/providers/__init__.py (hunk 14 above)
Comment (zas) [inline picard/coverart/providers/__init__.py]: Yes, replace by "must" in 45fc179

File: picard/ui/options/cover.py (hunk 15 above)
Comment (zas) [inline picard/ui/options/cover.py]: Nope, it is a leftover from previous versions, removed in c73f926

**Comment by zas** [comment]
@mineo : thank you for reviewing ;) please have a last look and tell me if we can merge it.

**Comment by zas** [comment]
Replaced by a much cleaner version : https://github.com/musicbrainz/picard/issues/416
